drivers/adc/adc_ifx_cat1.c

```diff
@@ -0,0 +1,271 @@
+#endif
+
+#define ADC_CAT1_EVENTS_MASK (CYHAL_ADC_EOS | CYHAL_ADC_ASYNC_READ_COMPLETE)
+
+#define ADC_CAT1_ACQUISITION_TIME_NS (1000u)
+#define ADC_CAT1_ADC_RESOLUTION      (12u)
+#define ADC_CAT1_REF_INTERNAL_MV     (1200u)
+
+struct ifx_cat1_adc_data {
+	struct adc_context ctx;
+	const struct device *dev;
+	cyhal_adc_t adc_obj;
+	cyhal_adc_channel_t adc_chan_obj[CY_SAR_SEQ_NUM_CHANNELS];
+	uint16_t *buffer;
+	uint16_t *repeat_buffer;
+	uint32_t channels;
+	uint32_t channels_mask;
+};
+
+struct ifx_cat1_adc_config {
+	uint8_t irq_priority;
+};
+
+static void _cyhal_adc_event_callback(void *callback_arg, cyhal_adc_event_t event)
+{
+	const struct device *dev = (const struct device *) callback_arg;
+	struct ifx_cat1_adc_data *data = dev->data;
+	uint32_t channels = data->channels;
+	int32_t result;
+	uint32_t channel_id;
+
+	while (channels != 0) {
+		channel_id = find_lsb_set(channels) - 1;
+		channels &= ~BIT(channel_id);
+
+		result = Cy_SAR_GetResult32(data->adc_chan_obj[channel_id].adc->base,
+					    data->adc_chan_obj[channel_id].channel_idx);
+		/* Legacy API for BWC. Convert from signed to unsigned by adding 0x800 to
+		 * convert the lowest signed 12-bit number to 0x0.
+		 */
+		*data->buffer = (uint16_t)(result + 0x800);
+		data->buffer++;
+	}
+
+	adc_context_on_sampling_done(&data->ctx, dev);
+
+	LOG_DBG("%s ISR triggered.", dev->name);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct ifx_cat1_adc_data *data = CONTAINER_OF(ctx, struct ifx_cat1_adc_data, ctx);
+
+	data->repeat_buffer = data->buffer;
+
+	Cy_SAR_StartConvert(data->adc_obj.base, CY_SAR_START_CONVERT_SINGLE_SHOT);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct ifx_cat1_adc_data *data = CONTAINER_OF(ctx, struct ifx_cat1_adc_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static int ifx_cat1_adc_channel_setup(const struct device *dev,
+				      const struct adc_channel_cfg *channel_cfg)
+{
+	struct ifx_cat1_adc_data *data = dev->data;
+	cy_rslt_t result;
+
+	cyhal_gpio_t vplus = CYHAL_GET_GPIO(_ADCSAR_PORT, channel_cfg->input_positive);
+	cyhal_gpio_t vminus = channel_cfg->differential ?
+			      CYHAL_GET_GPIO(_ADCSAR_PORT, channel_cfg->input_negative) :
+			      CYHAL_ADC_VNEG;
+
+	/* ADC channel configuration */
+	const cyhal_adc_channel_config_t channel_config = {
+		/* Disable averaging for channel */
+		.enable_averaging = false,
+		/* Minimum acquisition time set to 1us */
+		.min_acquisition_ns = ADC_CAT1_ACQUISITION_TIME_NS,
+		/* Sample channel when ADC performs a scan */
+		.enabled = true
+	};
+
+	/* Initialize a channel and configure it to scan the input pin(s). */
+	cyhal_adc_channel_free(&data->adc_chan_obj[channel_cfg->channel_id]);
+	result = cyhal_adc_channel_init_diff(&data->adc_chan_obj[channel_cfg->channel_id],
+					     &data->adc_obj, vplus, vminus, &channel_config);
```

Comment:
Why `gain`, `reference`, and `acquisition_time` from `channel_cfg` are ignored here? The driver should return `-EINVAL` if it cannot set the requested configuration.

Reply:
Done